### PR TITLE
Add inference api guide

### PR DESCRIPTION
**Pull request by steveganzy**

### CODE REVIEW

File: inference-explainers/inference_api.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+#### Individual request mandatory fields
+
+- **model_path:** Specifies the target model for the request
+- **tensors:** An array of JSON objects representing the input tensors to the model.
+
+#### Tensor mandatory fields
+
+- **data_type:** Specifies the data type of the tensor, chosen from DOUBLE, FLOAT, INT8, INT16, INT32, INT64.
+- **tensor_shape:** An array specifying the shape of the tensor.
+- **tensor_content:** A flattened representation of the tensor values.
```
Comment (akshaypundle) [inline inference-explainers/inference_api.md:89]: wonder if we should say "flattened row major representation .."

**Review by akshaypundle — CHANGES_REQUESTED** [review]

File: inference-explainers/inference_api.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+The Tensorflow and PyTorch runtimes are supported, each with a dedicated inference service sidecar binary. Currently, we offer the Tensorflow v2.14.0 sidecar binary and the PyTorch v2.1.1 sidecar binary.
+
+For Tensorflow, the SavedModel format is used while PyTorch uses the TorchScript format. We recommend that ad techs develop and train models in an environment in which the ML runtime version aligns with the intended sidecar.
+
+The maximum size for a single ML model is 2GB.
+
+## Model management
+
+ML models used by the inference service are fetched periodically from a linked cloud storage (e.g., [Google Cloud Storage buckets][2], [Amazon S3 buckets][3]) and made available for serving. Each model is uniquely identified by its path within the cloud storage. Ad techs can implement model versioning by structuring model storage paths to include version identifiers. For example, storing models at paths such as “pcvr_v1” and “pcvr_v2” distinguishes between the two versions.
```
[the remark this answers is not in the document]
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Thanks, I agree that this has a better flow of ideas. Incorporated.

File: inference-explainers/inference_api.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    {
+      "model_path": "pcvr_v2",
+      "checksum": "...",
+      "warm_up_batch_request_json": "..."
+    }
+  ]
+}
+```
+
+This model configuration file features a top-level array with each entry containing the metadata of the fetched models. The mandatory **model_path** field specifies the path to the model in the cloud storage. The optional **checksum** field is the SHA256 checksum of the model represented as a hexadecimal string.
```
[the remark this answers is not in the document]
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Done

File: inference-explainers/inference_api.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+1. Compute the SHA256 checksum for each individual model file.
+2. Arrange the file checksums in ascending order based on their file paths.
+3. Concatenate all the ordered file checksums into a single string.
+4. Compute the final SHA256 checksum on the concatenated checksum string.
+
+The listed steps are equivalent to the following bash command:
+
+```
+find <model_path> -type f -exec sha256sum {} \; | sort -k 2 | awk '{print $1}' | tr -d '\n' | sha256sum | awk '{print $1}'
```
[the remark this answers is not in the document]
Comment (steveganzy) [reply inference-explainers/inference_api.md:55]: We actually don't need to. This command already works for both cases. Because the find command checks the directory entries, it can figure out if the model path is a single file or a directory and performs a match only if a corresponding entry is found. We ran into the issue in our case because CPIO lacks the notion of directory.

More concretely, if model is a directory, model1 is a file or directory, "find model -type f" only finds model/file1 and model/file2 because in the current directory, there is an entry for "model" which is a directory. It then recursively traverses "model". "find model/ -type f" produces the same result. 

If model is a regular file. "find model -type f" finds the regular file only. In both cases, model1 is ignored.

[the remark this answers is not in the document]
Comment (steveganzy) [reply inference-explainers/inference_api.md:55]: Done

File: inference-explainers/inference_api.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+```
+
+The optional **warm_up_batch_request_json** is a JSON batch request used to warm up the model before any traffic is served. The format of this warm-up request uses the same format as the input to the runInference JavaScript function, which is described later in this document.
```
[the remark this answers is not in the document]
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Done

File: inference-explainers/inference_api.md (hunk 1 above)
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Done

File: inference-explainers/inference_api.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    {
+      "model_path": "pcvr_v2",
+      "checksum": "...",
+      "warm_up_batch_request_json": "..."
+    }
+  ]
+}
+```
+
+This model configuration file features a top-level array with each entry containing the metadata of the fetched models. The mandatory **model_path** field specifies the path to the model in the cloud storage. If the specified model path points to a directory, it needs to end with a "/" suffix. This suffix indicates that the entire directory will be used for model registration. Without the "/" suffix, an exact path match is expected, and only the specified file will be registered as a model. In the example above, both model paths refer to single model files. If the paths were to refer to directories instead, they would need to be modified to "pcvr_1/" and "pcvr_2/". The optional **checksum** field is the SHA256 checksum of the model represented as a hexadecimal string.
```
Comment (mihnjong-l) [inline inference-explainers/inference_api.md:43]: It's a bit confusing.

Can we give two concrete examples?

For PyTorch, can we use the path like pcvr_v1/model.pt to clearly show that model.pt is a single model file that we're using?

For TensorFlow, a model consists of multiple files and sub-directories, so we can show the example of "pcvr_v1/" where saved_model.pb and other files are in the directory.

Comment (mihnjong-l) [inline inference-explainers/inference_api.md:43]: Usually, you use `field` instead of **field**.

File: inference-explainers/inference_api.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+- **Batch with all successful outputs:** The top-level array contains multiple response objects, each with the **model_path** and **tensor** fields populated which indicates successful inference for each input in the batch:
+
+```json
+{
+  "response": [
+    {
+      "model_path": "pcvr_1",
+      "tensors": [
+        {
+          "tensor_name": "StatefulPartitionedCall:0",
```
Comment (mihnjong-l) [inline inference-explainers/inference_api.md:162]: Nit: Let's use "PartitionedCall:0" assuming that the model is frozen.

**Review by mihnjong-l — CHANGES_REQUESTED** [review]
PTAL

File: inference-explainers/inference_api.md (hunk 6 above)
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Done

File: inference-explainers/inference_api.md (hunk 7 above)
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Done

File: inference-explainers/inference_api.md (hunk 6 above)
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Thanks done

File: inference-explainers/inference_api.md [hunk 8]
```diff
@@ -19,43 +19,54 @@ The maximum size for a single ML model is 2GB.
 
 ML models used by the inference service are fetched periodically from a linked cloud storage (e.g., [Google Cloud Storage buckets][2], [Amazon S3 buckets][3]) and made available for serving. Each model is uniquely identified by its path within the cloud storage. To determine the models to load, the inference service looks for a JSON model configuration file in the cloud bucket. Ad techs should maintain a model configuration file in JSON format within the same cloud storage as the models. The B&A service periodically checks the configuration file (according to period configurable by ad techs) for any changes and triggers the loading of new models into the inference service sidecar’s memory as needed.
 
-Ad techs can implement model versioning by structuring model storage paths to include version identifiers. For example, storing models at paths such as “pcvr_v1” and “pcvr_v2” distinguishes between the two versions.
+Ad techs can implement model versioning by structuring model storage paths to include version identifiers. For example, storing models under the direcotries such as “pcvr_v1/” and “pcvr_v2/” distinguishes between the two versions.
```
Comment (mihnjong-l) [inline inference-explainers/inference_api.md:22]: s/direcotries/directories

**Review by mihnjong-l — APPROVED** [review]
LGTM. Please fix a typo before merging.

File: inference-explainers/inference_api.md (hunk 8 above)
Comment (steveganzy) [reply inference-explainers/inference_api.md]: Done

**Review by mihnjong-l — APPROVED** [review]
Looks great!

File: inference-explainers/inference_api.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+For Tensorflow, the SavedModel format is used while PyTorch uses the TorchScript format. We recommend that ad techs develop and train models in an environment in which the ML runtime version aligns with the intended sidecar.
+
+The maximum size for a single ML model is 2GB.
+
+## Model management
+
+ML models used by the inference service are fetched periodically from a linked cloud storage (e.g., [Google Cloud Storage buckets][2], [Amazon S3 buckets][3]) and made available for serving. Each model is uniquely identified by its path within the cloud storage. To determine the models to load, the inference service looks for a JSON model configuration file in the cloud bucket. Ad techs should maintain a model configuration file in JSON format within the same cloud storage as the models. The B&A service periodically checks the configuration file (according to period configurable by ad techs) for any changes and triggers the loading of new models into the inference service sidecar’s memory as needed.
```
Comment (akshaypundle) [inline inference-explainers/inference_api.md:20]: one small suggestion: can we also mention how the system knows that the JSON model config file is? Maybe mention the config setting for this ...

**Review by akshaypundle — APPROVED** [review]
